File: backend/src/main/java/org/example/backend/domain/log/service/AbstractLogService.java
package org.example.backend.domain.log.service;

import org.example.backend.global.exception.BusinessException;
import org.example.backend.global.exception.ErrorCode;
import org.springframework.data.jpa.repository.JpaRepository;

import java.util.List;
import java.util.stream.Collectors;

// 로그 서비스 공통 추상 클래스 aquariumlog, fishlog 중복코드 제거
public abstract class AbstractLogService<Entity, RequestDto, ResponseDto, ParentEntity> {

    protected abstract JpaRepository<Entity, Long> getLogRepository();
    protected abstract JpaRepository<ParentEntity, Long> getParentRepository();
    protected abstract ErrorCode getLogNotFoundErrorCode();
    protected abstract ErrorCode getParentNotFoundErrorCode();
    protected abstract Entity createEntity(RequestDto requestDto, ParentEntity parent);
    protected abstract ResponseDto convertToResponseDto(Entity entity);
    protected abstract void updateEntity(Entity entity, RequestDto requestDto, ParentEntity parent);
    protected abstract List<Entity> findByParentId(Long parentId);

    // 로그생성
    public ResponseDto createLog(RequestDto requestDto, Long parentId) {
        ParentEntity parent = getParentRepository().findById(parentId)
                .orElseThrow(() -> new BusinessException(getParentNotFoundErrorCode()));
        
        Entity entity = createEntity(requestDto, parent);
        Entity savedEntity = getLogRepository().save(entity);
        return convertToResponseDto(savedEntity);
    }

    // 로그목록 조회
    public List<ResponseDto> getLogsByParentId(Long parentId) {
        return findByParentId(parentId).stream()
                .map(this::convertToResponseDto)
                .collect(Collectors.toList());
    }

    //로그 수정
    public ResponseDto updateLog(Long logId, RequestDto requestDto, Long parentId) {
        Entity entity = getLogRepository().findById(logId)
                .orElseThrow(() -> new BusinessException(getLogNotFoundErrorCode()));

        ParentEntity parent = getParentRepository().findById(parentId)
                .orElseThrow(() -> new BusinessException(getParentNotFoundErrorCode()));

        updateEntity(entity, requestDto, parent);
        return convertToResponseDto(entity);
    }

    //로그 삭제 
    public void deleteLog(Long logId) {
        Entity entity = getLogRepository().findById(logId)
                .orElseThrow(() -> new BusinessException(getLogNotFoundErrorCode()));
        getLogRepository().delete(entity);
    }
}

File: backend/src/main/java/org/example/backend/domain/log/service/AbstractLogService.kt
package org.example.backend.domain.log.service

import org.example.backend.global.exception.BusinessException
import org.example.backend.global.exception.ErrorCode
import org.springframework.data.jpa.repository.JpaRepository

/**
 * 로그 서비스 공통 추상 클래스
 * AquariumLog, FishLog 등 다양한 로그 서비스의 공통 기능을 제공
 * 
 * @param Entity 로그 엔티티 타입
 * @param RequestDto 요청 DTO 타입
 * @param ResponseDto 응답 DTO 타입
 * @param ParentEntity 부모 엔티티 타입 (Aquarium, Fish 등)
 */
abstract class AbstractLogService<Entity, RequestDto, ResponseDto, ParentEntity> {
    
    protected abstract fun getLogRepository(): JpaRepository<Entity, Long>
    protected abstract fun getParentRepository(): JpaRepository<ParentEntity, Long>
    protected abstract fun getLogNotFoundErrorCode(): ErrorCode
    protected abstract fun getParentNotFoundErrorCode(): ErrorCode
    protected abstract fun createEntity(requestDto: RequestDto, parent: ParentEntity): Entity
    protected abstract fun convertToResponseDto(entity: Entity): ResponseDto
    protected abstract fun updateEntity(entity: Entity, requestDto: RequestDto, parent: ParentEntity)
    protected abstract fun findByParentId(parentId: Long): List<Entity>

    /**
     * 로그 생성
     * @param requestDto 요청 DTO
     * @param parentId 부모 엔티티 ID
     * @return 생성된 로그의 응답 DTO
     */
    fun createLog(requestDto: RequestDto, parentId: Long): ResponseDto {
        val parent = getParentRepository().findById(parentId)
            .orElseThrow { BusinessException(getParentNotFoundErrorCode()) }
        
        val entity = createEntity(requestDto, parent)
        // 코틀린의 제네릭 타입 추론 문제를 해결하기 위한 타입 캐스팅
        // JpaRepository.save()는 <S extends T> S save(S entity) 형태이므로 타입 캐스팅 필요
        // 임시 코드: Any로 캐스팅하여 타입 체크 우회
        @Suppress("UNCHECKED_CAST")
        val savedEntity: Entity = (getLogRepository() as JpaRepository<Any, Long>).save(entity as Any) as Entity
        return convertToResponseDto(savedEntity)
    }

    /**
     * 부모 ID로 로그 목록 조회
     * @param parentId 부모 엔티티 ID
     * @return 로그 응답 DTO 목록
     */
    fun getLogsByParentId(parentId: Long): List<ResponseDto> {
        return findByParentId(parentId).map { convertToResponseDto(it) }
    }

    /**
     * 로그 수정
     * @param logId 로그 ID
     * @param requestDto 요청 DTO
     * @param parentId 부모 엔티티 ID
     * @return 수정된 로그의 응답 DTO
     */
    fun updateLog(logId: Long, requestDto: RequestDto, parentId: Long): ResponseDto {
        val entity = getLogRepository().findById(logId)
            .orElseThrow { BusinessException(getLogNotFoundErrorCode()) }

        val parent = getParentRepository().findById(parentId)
            .orElseThrow { BusinessException(getParentNotFoundErrorCode()) }

        updateEntity(entity, requestDto, parent)
        return convertToResponseDto(entity)
    }

    /**
     * 로그 삭제
     * 하위 클래스에서 오버라이드 가능하도록 open으로 선언
     * @param logId 로그 ID
     */
    open fun deleteLog(logId: Long) {
        val entity = getLogRepository().findById(logId)
            .orElseThrow { BusinessException(getLogNotFoundErrorCode()) }
        // 코틀린의 제네릭 타입 추론 문제를 해결하기 위한 타입 캐스팅
        // 임시 코드: Any로 캐스팅하여 타입 체크 우회
        @Suppress("UNCHECKED_CAST")
        (getLogRepository() as JpaRepository<Any, Long>).delete(entity as Any)
    }
}
